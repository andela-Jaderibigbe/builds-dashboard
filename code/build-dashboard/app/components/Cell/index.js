import React from 'react';
import PropTypes from 'prop-types';
import styles from './Cell.scss';

const Cell = ({ className, children, isHidden }) => {
  className = `${styles.cell} ${className}`;
  return <div className={className}>{!isHidden ? children : null}</div>;
};

Cell.defaultProps = {
  isHidden: false,
};

Cell.propTypes = {
  className: PropTypes.string,
  isHidden: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default Cell;
