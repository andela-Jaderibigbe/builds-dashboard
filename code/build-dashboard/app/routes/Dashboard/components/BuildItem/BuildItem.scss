@import 'theme/variables';

.tableRow {
  width: 98%;
  margin: 1%;
}

.changeListCol {
  width: 19%;
  text-align: left;
}

.ownerCol {
  text-align: center;
  width: 9%;
}

.timeCol {
  text-align: center;
  width: 14%;
}

.stateCol {
  text-align: center;
  width: 14%;
}

.statusCol {
  text-align: center;
  width: 11%;
}

.buildName {
  cursor: pointer;
}

.buildName:hover {
  text-decoration: underline;
}

.buildIcon {
  i {
    margin-right: 15px;
  }
}

.isRunning {
  color: $blue;
  border: 1px solid $blue;
  background-image: linear-gradient(
    to right,
    rgba($blue, 0.1) 0%,
    transparent 50%
  );
}

.isPending {
  color: $gray;
  border: 1px solid $gray;
  background-image: linear-gradient(
    to right,
    rgba($gray, 0.1) 0%,
    transparent 50%
  );
}

.isRejected {
  color: $red;
  border: 1px solid $red;
  background-image: linear-gradient(
    to right,
    rgba($red, 0.1) 0%,
    transparent 50%
  );
}

.isSuccess {
  color: $green;
  border: 1px solid $green;
  background-image: linear-gradient(
    to right,
    rgba($green, 0.1) 0%,
    transparent 50%
  );
}
