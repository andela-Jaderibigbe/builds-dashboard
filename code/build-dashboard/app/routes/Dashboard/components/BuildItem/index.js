import React from 'react';
import PropTypes from 'prop-types';
import ExpandableRow from 'HOCS/ExpandableRow';
import Cell from 'components/Cell';
import TableRow from 'components/TableRow';
import Progress from 'components/Progress';
import Stats from '../Stats';
import { parseDatetime } from 'utils';
import { stateToDisplay, hasDrop } from '../../selector';
import styles from './BuildItem.scss';

const RowWithDetails = ExpandableRow(TableRow, Stats);

const getIconClassName = state => {
  switch (state) {
    case 'running':
      return 'fa fa-hourglass-half';
    case 'pending':
      return 'fa fa-hourglass-start';
    default:
      return 'fa fa-hourglass-end';
  }
};

const getRowStatus = state => {
  switch (state) {
    case 'running':
      return styles.isRunning;
    case 'pending':
      return styles.isPending;
    case 'rejected':
      return styles.isRejected;
    default:
      return styles.isSuccess;
  }
};

const BuildItem = props => {
  const {
    _id,
    user,
    build_name,
    state,
    progress: { metrics, build, unit_test, functional_test },
    time_started,
    stats,
    isSelected,
    onSelectItem,
  } = props;

  return (
    <RowWithDetails
      isSelected={isSelected}
      stats={stats}
      progress={props.progress}
      className={`${styles.tableRow} ${getRowStatus(state)}`}
      state={state}
    >
      <Cell className={styles.changeListCol}>
        <span className={styles.buildIcon}>
          <i className={getIconClassName(state)} />
          <span className={styles.buildName} onClick={() => onSelectItem(_id)}>
            {build_name}
          </span>
        </span>
      </Cell>
      <Cell className={styles.ownerCol}>{user}</Cell>
      <Cell className={styles.timeCol}>{parseDatetime(time_started)}</Cell>
      <Cell className={styles.stateCol}>{stateToDisplay(state)}</Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={metrics}
          progressState={state}
          hasError={hasDrop(stats['metrics'] || {})}
        />
      </Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={build}
          progressState={state}
          hasError={hasDrop(stats['build'] || {})}
        />
      </Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={unit_test}
          progressState={state}
          hasError={hasDrop(stats['unit_test'] || {})}
        />
      </Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={functional_test}
          progressState={state}
          hasError={hasDrop(stats['functional_test'] || {})}
        />
      </Cell>
    </RowWithDetails>
  );
};

BuildItem.propTypes = {
  _id: PropTypes.number,
  user: PropTypes.string,
  build_name: PropTypes.string,
  state: PropTypes.string,
  progress: PropTypes.object,
  time_started: PropTypes.string,
  stats: PropTypes.object,
  isSelected: PropTypes.bool,
  onSelectItem: PropTypes.func,
};

export default BuildItem;
