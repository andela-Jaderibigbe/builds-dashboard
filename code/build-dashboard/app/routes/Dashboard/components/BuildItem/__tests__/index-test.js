import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { expect as chaiExpect } from 'chai';
import Data from 'data.json';

import Cell from 'components/Cell';
import BuildItem from '../';

describe('<BuildItem />', () => {
  const props = Data[2];

  it('should render correctly', () => {
    const tree = renderer.create(<BuildItem {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have correct number of columns', () => {
    const wrapper = shallow(<BuildItem {...props} />);
    chaiExpect(wrapper).to.exist;
    chaiExpect(wrapper.find(Cell)).to.have.length(8);
  });

  it('should contain build name', () => {
    const wrapper = shallow(<BuildItem {...props} />);

    chaiExpect(wrapper.find('.buildName').text()).to.equal(props.build_name);
  });
});
