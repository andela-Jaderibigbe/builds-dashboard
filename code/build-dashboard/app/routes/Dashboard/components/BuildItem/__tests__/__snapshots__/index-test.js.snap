// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BuildItem /> should render correctly 1`] = `
<div
  className="tableRow tableRow isRejected"
>
  <div
    className="cell changeListCol"
  >
    <span
      className="buildIcon"
    >
      <i
        className="fa fa-hourglass-end"
      />
      <span
        className="buildName"
        onClick={[Function]}
      >
        432461
      </span>
    </span>
  </div>
  <div
    className="cell ownerCol"
  >
    samy
  </div>
  <div
    className="cell timeCol"
  >
    10/13/2017 02:48 pm
  </div>
  <div
    className="cell stateCol"
  >
    Rejected
  </div>
  <div
    className="cell statusCol"
  >
    <div
      className="progressBar"
    >
      <span
        className="isError"
        style={
          Object {
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="cell statusCol"
  >
    <div
      className="progressBar"
    >
      <span
        className="isSuccess"
        style={
          Object {
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="cell statusCol"
  >
    <div
      className="progressBar"
    >
      <span
        className="isSuccess"
        style={
          Object {
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="cell statusCol"
  >
    <div
      className="progressBar"
    >
      <span
        className="isSuccess"
        style={
          Object {
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </div>
  </div>
</div>
`;
